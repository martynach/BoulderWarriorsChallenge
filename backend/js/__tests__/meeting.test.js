const Meeting = require('./../Meeting');
const meeting = new Meeting();
const path = require("path");
const filepath = path.join( __dirname, './test_data/meetings.json');
meeting.filepath = filepath;

const fs = require('fs');
const promisify = require('./../promisify');



test('Get all meetings', async () => {
    expect.assertions(1);
    const actual = await meeting.getAllMeetings();
    expect(actual).toMatchSnapshot();
});

test('Get results for meeting with id=1', async () => {
    expect.assertions(1);
    const actual = await meeting.getResults(1);
    expect(actual).toMatchSnapshot();
});

test('Get results for no existing meeting in the future wth id=2', async () => {
    expect.assertions(1);
    const actual = await meeting.getResults(2);
    expect(actual).toHaveLength(0);
});


//**************************
describe('Tests for adding meeting', () => {
    const tmpFilepath = path.join(__dirname, './test_data/meetings_tmp.json');

    beforeAll(() => {
        meeting.filepath = tmpFilepath;
    });

    beforeEach(async () => {
        await promisify(fs.copyFile, filepath, tmpFilepath);

    });

    // //TODO
    // afterEach(async () => {
    //     await promisify(fs.ulink, tmpFilepath);
    // });

    afterAll(() => {
        meeting.filepath = filepath;
    });

    test('addNewMeeting adding single future event', async () => {
        expect.assertions(3);

        const newEvent = { name: 'new meeting', date: '5.06.2018', numOfBoulders: 10 };
        const addingResult = await meeting.addNewMeeting(newEvent);
        expect(addingResult).toBeTruthy();

        const meetings = await meeting.getAllMeetings();
        expect(meetings).toMatchSnapshot();

        const tmpFileContent = await promisify(fs.readFile, tmpFilepath, 'utf8');
        expect(tmpFileContent).toMatchSnapshot();
    });

    test('addNewMeeting adding single future event with some players', async () => {
        expect.assertions(3);

        const newEvent = { name: 'new meeting', date: '5.06.2018', numOfBoulders: 10, players: [1,2] };
        const addingResult = await meeting.addNewMeeting(newEvent);
        expect(addingResult).toBeTruthy();

        const meetings = await meeting.getAllMeetings();
        expect(meetings).toMatchSnapshot();

        const tmpFileContent = await promisify(fs.readFile, tmpFilepath, 'utf8');
        expect(tmpFileContent).toMatchSnapshot();
    });

    test('addNewMeeting adding single future event with incorrect players', async () => {
        expect.assertions(1);

        const newEvent = { name: 'new meeting', date: '5.06.2018', numOfBoulders: 10, players: [1,2, 200] };
        const addingResult = await meeting.addNewMeeting(newEvent);

        expect(addingResult).toBeFalsy();

    });

});