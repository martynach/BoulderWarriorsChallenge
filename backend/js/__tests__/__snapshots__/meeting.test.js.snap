// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get all meetings 1`] = `
Array [
  Object {
    "date": "13.01.2018",
    "id": 1,
    "name": "Pierwsze starcie wojowników",
    "numOfBoulders": 5,
    "players": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
    ],
    "results": Array [
      Object {
        "bonus": 3,
        "playerID": 1,
        "top": 2,
      },
      Object {
        "bonus": 5,
        "playerID": 2,
        "top": 3,
      },
      Object {
        "bonus": 5,
        "playerID": 3,
        "top": 5,
      },
      Object {
        "bonus": 5,
        "playerID": 4,
        "top": 4,
      },
      Object {
        "bonus": 4,
        "playerID": 5,
        "top": 3,
      },
      Object {
        "bonus": 2,
        "playerID": 6,
        "top": 1,
      },
    ],
  },
  Object {
    "date": "13.05.2018",
    "id": 2,
    "name": "Boulder Warriors Challenge - długa walka",
    "numOfBoulders": 40,
    "players": Array [
      1,
      9,
    ],
    "results": Array [],
  },
]
`;

exports[`Get results for meeting with id=1 1`] = `
Array [
  Object {
    "bonus": 5,
    "firstname": "Wojciech",
    "gender": "m",
    "lastname": "Bańczyk",
    "top": 5,
  },
  Object {
    "bonus": 5,
    "firstname": "Karolina",
    "gender": "f",
    "lastname": "Kozakiewicz",
    "top": 4,
  },
  Object {
    "bonus": 5,
    "firstname": "Michał",
    "gender": "m",
    "lastname": "Nowak",
    "top": 3,
  },
  Object {
    "bonus": 4,
    "firstname": "Damian",
    "gender": "m",
    "lastname": "Strojny",
    "top": 3,
  },
  Object {
    "bonus": 3,
    "firstname": "Edyta",
    "gender": "f",
    "lastname": "Tokarczyk",
    "top": 2,
  },
  Object {
    "bonus": 2,
    "firstname": "Jakub",
    "gender": "m",
    "lastname": "Bogucki",
    "top": 1,
  },
]
`;

exports[`Tests for adding meeting addNewMeeting adding single future meeting 1`] = `
Array [
  Object {
    "date": "13.01.2018",
    "id": 1,
    "name": "Pierwsze starcie wojowników",
    "numOfBoulders": 5,
    "players": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
    ],
    "results": Array [
      Object {
        "bonus": 3,
        "playerID": 1,
        "top": 2,
      },
      Object {
        "bonus": 5,
        "playerID": 2,
        "top": 3,
      },
      Object {
        "bonus": 5,
        "playerID": 3,
        "top": 5,
      },
      Object {
        "bonus": 5,
        "playerID": 4,
        "top": 4,
      },
      Object {
        "bonus": 4,
        "playerID": 5,
        "top": 3,
      },
      Object {
        "bonus": 2,
        "playerID": 6,
        "top": 1,
      },
    ],
  },
  Object {
    "date": "13.05.2018",
    "id": 2,
    "name": "Boulder Warriors Challenge - długa walka",
    "numOfBoulders": 40,
    "players": Array [
      1,
      9,
    ],
    "results": Array [],
  },
  Object {
    "date": "5.06.2018",
    "id": 3,
    "name": "new meeting",
    "numOfBoulders": 10,
    "players": Array [],
    "results": Array [],
  },
]
`;

exports[`Tests for adding meeting addNewMeeting adding single future meeting 2`] = `"[{\\"id\\":1,\\"name\\":\\"Pierwsze starcie wojowników\\",\\"date\\":\\"13.01.2018\\",\\"numOfBoulders\\":5,\\"players\\":[1,2,3,4,5,6,7],\\"results\\":[{\\"playerID\\":1,\\"top\\":2,\\"bonus\\":3},{\\"playerID\\":2,\\"top\\":3,\\"bonus\\":5},{\\"playerID\\":3,\\"top\\":5,\\"bonus\\":5},{\\"playerID\\":4,\\"top\\":4,\\"bonus\\":5},{\\"playerID\\":5,\\"top\\":3,\\"bonus\\":4},{\\"playerID\\":6,\\"top\\":1,\\"bonus\\":2}]},{\\"id\\":2,\\"name\\":\\"Boulder Warriors Challenge - długa walka\\",\\"date\\":\\"13.05.2018\\",\\"numOfBoulders\\":40,\\"players\\":[1,9],\\"results\\":[]},{\\"id\\":3,\\"name\\":\\"new meeting\\",\\"date\\":\\"5.06.2018\\",\\"numOfBoulders\\":10,\\"players\\":[],\\"results\\":[]}]"`;

exports[`Tests for adding meeting addNewMeeting adding single meeting with some players 1`] = `
Array [
  Object {
    "date": "13.01.2018",
    "id": 1,
    "name": "Pierwsze starcie wojowników",
    "numOfBoulders": 5,
    "players": Array [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
    ],
    "results": Array [
      Object {
        "bonus": 3,
        "playerID": 1,
        "top": 2,
      },
      Object {
        "bonus": 5,
        "playerID": 2,
        "top": 3,
      },
      Object {
        "bonus": 5,
        "playerID": 3,
        "top": 5,
      },
      Object {
        "bonus": 5,
        "playerID": 4,
        "top": 4,
      },
      Object {
        "bonus": 4,
        "playerID": 5,
        "top": 3,
      },
      Object {
        "bonus": 2,
        "playerID": 6,
        "top": 1,
      },
    ],
  },
  Object {
    "date": "13.05.2018",
    "id": 2,
    "name": "Boulder Warriors Challenge - długa walka",
    "numOfBoulders": 40,
    "players": Array [
      1,
      9,
    ],
    "results": Array [],
  },
  Object {
    "date": "5.06.2018",
    "id": 3,
    "name": "new meeting",
    "numOfBoulders": 10,
    "players": Array [],
    "results": Array [],
  },
  Object {
    "date": "5.06.2018",
    "id": 4,
    "name": "new meeting",
    "numOfBoulders": 10,
    "players": Array [
      1,
      2,
    ],
    "results": Array [],
  },
]
`;

exports[`Tests for adding meeting addNewMeeting adding single meeting with some players 2`] = `"[{\\"id\\":1,\\"name\\":\\"Pierwsze starcie wojowników\\",\\"date\\":\\"13.01.2018\\",\\"numOfBoulders\\":5,\\"players\\":[1,2,3,4,5,6,7],\\"results\\":[{\\"playerID\\":1,\\"top\\":2,\\"bonus\\":3},{\\"playerID\\":2,\\"top\\":3,\\"bonus\\":5},{\\"playerID\\":3,\\"top\\":5,\\"bonus\\":5},{\\"playerID\\":4,\\"top\\":4,\\"bonus\\":5},{\\"playerID\\":5,\\"top\\":3,\\"bonus\\":4},{\\"playerID\\":6,\\"top\\":1,\\"bonus\\":2}]},{\\"id\\":2,\\"name\\":\\"Boulder Warriors Challenge - długa walka\\",\\"date\\":\\"13.05.2018\\",\\"numOfBoulders\\":40,\\"players\\":[1,9],\\"results\\":[]},{\\"id\\":3,\\"name\\":\\"new meeting\\",\\"date\\":\\"5.06.2018\\",\\"numOfBoulders\\":10,\\"players\\":[],\\"results\\":[]},{\\"id\\":4,\\"name\\":\\"new meeting\\",\\"date\\":\\"5.06.2018\\",\\"numOfBoulders\\":10,\\"players\\":[1,2],\\"results\\":[]}]"`;
